### PR TITLE
Add option to map API tracking parameter to custom variable in log importer script.

### PR DESCRIPTION
**Pull request by diosmosis**
This pull request adds two options to import_logs.py that move tracking API parameters to custom variables. This is for a PRO request to allow not tracking user IDs from `cs-username` using uid and instead track via a custom variable. To do this, you would specify `--api-arg-to-visit-cvar="uid=User Name"`. The option references tracking API parameters so it can be used w/ any format.

I think using API parameters might be limiting however. After thinking about it, it might be better to make it use regex groups. Then we can add two more options, `--ignore-fields=userid`, `--w3c-field-regex=my-custom-field=myregex` so users can import any field they want into a custom variable.

Includes change to ImportLogs system test and fix for import log custom variable handling (old behavior was to not set multiple custom variables if they were set already instead of simply adding custom variables).

@mattab @quba Can you let me know what you think?

### CODE REVIEW

File: misc/log-analytics/import_logs.py [hunk 1]
```diff
@@ -1446,8 +1473,40 @@ def _get_hit_args(self, hit):
...
+
         return args
 
+    def _get_api_args_custom_variables(self, custom_vars, api_args_to_cvars_map, request_args):
+        """
+        Handles the --api-arg-to-...-cvar options by moving API request query parameters to
+        a custom variables data structure. The data structure is returned and can be used in
+        the cvar and _cvar query parameters.
+        """
+        # TODO: test not overwriting existing 'cvars'...
```
Comment (mattab) [inline misc/log-analytics/import_logs.py]: a TODO left here in case you missed it

**Comment by mattab** [comment]
it's a nice, clean re-usable solution. I like it! 

@quba did you have a chance to test it and can confirm the feature?

**Comment by diosmosis** [comment]
@mattab What did you think about the second paragraph?

**Comment by mattab** [comment]
> I think using API parameters might be limiting however. After thinking about it, it might be better to make it use regex groups. Then we can add two more options, --ignore-fields=userid, --w3c-field-regex=my-custom-field=myregex so users can import any field they want into a custom variable.

Yes, it could be useful to have this ability but maybe we wait for some use cases? AFAIK it would not invalidate this PR and new options so it could be done later
maybe create a separate issue?

**Comment by diosmosis** [comment]
It wouldn't invalidate it, but it would be redundant since you could use the other options to do what's in this use case.

**Comment by mattab** [comment]
What would be equivalent to `--api-arg-to-visit-cvar="uid=User Name"` with these options?

**Comment by diosmosis** [comment]
`--regex-group-to-visit-cvar="userid=User Name"`

**Comment by mattab** [comment]
ok looks nice! btw is there an option to list the regex groups available for a given log file?

**Comment by diosmosis** [comment]
It would be for a format, I was planning on adding the info to the help documentation.

**Comment by diosmosis** [comment]
> ok looks nice!

I'll make the changes, please don't merge until I finish them.

**Comment by diosmosis** [comment]
@mattab I've modified the pull request, new options now include:

```
--regex-group-to-visit-cvar
--regex-group-to-page-cvar
--ignore-fields
--w3c-field-regex
```

First 3 are for @quba's use case. Last one can be combined w/ other ones to track any w3c log field that the importer doesn't natively recognize in a custom variable.

Let me know what you think.

**Comment by quba** [comment]
@diosmosis how can I choose which custom variable will be used?

**Comment by diosmosis** [comment]
You mean which index? Or what the custom variable name to use? For name, --regex-group-to-visit-cvar="userid=User Name". I don't think index matters.

**Comment by mattab** [comment]
after a quick review, looks good to me! :+1:

**Comment by mattab** [comment]
@diosmosis in the morning i'm thinking, maybe we could add a test case in python for these new parameters? they can be difficult to grasp and this would ensure we don't regress without noticing.

**Comment by diosmosis** [comment]
There are tests in ImportLogsTest.php, do you mean add more tests to log-analytics/tests/*?

**Comment by mattab** [comment]
> do you mean add more tests to log-analytics/tests/*?

yes

**Pull request merged**